hatchet/readers/hpctoolkit_reader_latest.py

```diff
@@ -0,0 +1,421 @@
+            profile_db = file.read(szProfileInfos)
+            (pProfiles,) = safe_unpack("<Q", profile_db, 0)
+
+            (nValues, pValues, nCtxs, _, pCtxIndices) = safe_unpack(
+                "<QQLLQ", profile_db, pProfiles - pProfileInfos
+            )
+
+            file.seek(pCtxIndices)
+            formatCtxs = "<LQ"
+            cct_sub_db = file.read(nCtxs * struct.calcsize(formatCtxs))
+
+            file.seek(pValues)
+            formatValues = "<Hd"
+            values_sub_db = file.read(nValues * struct.calcsize(formatValues))
+
+            for i in range(nCtxs):
+                (ctxId, startIndex) = safe_unpack("<LQ", cct_sub_db, 0, i)
+                cct_id = ctxId
+
+                if i == nCtxs - 1:
+                    end_index = nValues
+                else:
+                    (_, end_index) = safe_unpack(formatCtxs, cct_sub_db, 0, i + 1)
+
+                self._summary_profile[cct_id] = {}
+
+                for j in range(startIndex, end_index):
+                    (metricId, value) = safe_unpack("<Hd", values_sub_db, 0, j)
+
+                    if metricId in self._metric_descriptions:
+                        self._summary_profile[cct_id][
+                            self._metric_descriptions[metricId]
+                        ] = value
+
+                        if self._metric_descriptions[metricId].endswith("(i)"):
+                            self._inclusive_metrics[metricId] = (
+                                self._metric_descriptions[metricId]
+                            )
+                        else:
+                            self._exclusive_metrics[metricId] = (
+                                self._metric_descriptions[metricId]
+                            )
+
+    def _read_cct(
+        self,
+    ) -> None:
+        hatchet_roots = []
+
+        with open(self._meta_file, "rb") as file:
+            meta_db = file.read()
+
+        (pContext,) = safe_unpack("<Q", meta_db, FILE_HEADER_OFFSET + 7 * 8)
+        (pEntryPoints, nEntryPoints, szEntryPoint) = safe_unpack(
+            "<QHB", meta_db, pContext
+        )
+
+        for i in range(nEntryPoints):
+            (szChildren, pChildren, ctxId, entryPoint) = safe_unpack(
+                "<QQLH",
+                meta_db,
+                pEntryPoints,
+                i,
+                szEntryPoint,
+            )
+
+            if entryPoint != 1:
+                continue
+
+            frame = {
+                "id": ctxId,
+                "type": "entry",
+                "name": entryPoint,
+                "depth": 0,
+            }
+            entry_node = Node(Frame(frame), parent=None, hnid=ctxId)
+
+            hatchet_roots.append(entry_node)
+
+            entry_value = {"node": entry_node, "name": "entry"}
+
+            if ctxId in self._summary_profile:
+                entry_value.update(self._summary_profile[ctxId])
+
+            self._metrics_table.append(entry_value)
```

Comment:
Could any of this code for creating the `Node` objects and updating `_metrics_table` be abstracted so it could be reused in both `_read_cct` and `_parse_context`? It would help make the code more maintainable in the future.

Reply:
@ilumsden check my latest commit, this is fixed